networkx/algorithms/centrality/rumor.py

```diff
@@ -0,0 +1,126 @@
+import math
+
+import networkx as nx
+from networkx.utils.decorators import not_implemented_for
+
+__all__ = ["rumor_centrality"]
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def rumor_centrality(G: nx.Graph):
+    """Compute the rumor centrality for nodes. Rumor centrality is a rumor source detector which uses
+    the network's topology to estimate which (single) vertex is the most likely source of the rumor. In
+    some cases (cf. [1]) it is the maximum likelihood estimator.
+
+    Parameters
+    ----------
+    G : graph
+      A networkx graph
+
+    Returns
+    -------
+    nodes : dictionary
+       Dictionary of nodes with degree centrality as the value.
+
+    Raises
+    ------
+    NetworkXPointlessConcept
+        If the graph G is the null graph.
+
+    NetworkXNotImplemented
+        If the graph is a multigraph, directed, or contains cycles.
+
+    Notes
+    -----
+    Rumor centrality was introduced by Shah & Zaman [1]_ for rumor source detection.
+
+    .. [1] Shah, Devavrat, and Tauhid Zaman. "Rumor centrality: a universal source detector." In Proceedings of the
+    12th ACM SIGMETRICS/PERFORMANCE joint international conference on Measurement and Modeling of Computer Systems,
+    pp. 199-210. 2012.
+
+    """
+
+    if len(G) == 0:
+        raise nx.NetworkXPointlessConcept(
+            "cannot compute centrality for the null graph"
+        )
+
+    if not nx.is_tree(G):
+        raise nx.NetworkXNotImplemented(
+            "rumor centrality does not work for graphs with cycles"
+        )
+
+    # Create copy to be able to use node properties
+    H = G.copy()
+    set_visited_to_false(H)
```

Comment:
It is likely to be much faster and probably more memory efficient if you store the node properties as dicts keyed by node to the value of that property rather than store them on `G`. It is also less likely to overwrite any user attributes.

For example, `visited` can be a set. Add nodes to the set when they are visited. check if they have been visited by `node in visited`.

Reply:
Thank you! I have implemented this as requested.